tests/driver_l3gxxxx/main.c

```diff
@@ -0,0 +1,280 @@
+    msg_t msg = { .content.value = L3GXXXX_INT_DATA_READY };
+    msg_send(&msg, p_main);
+}
+#endif
+
+#if IS_USED(MODULE_L3GXXXX_IRQ_EVENT)
+int handle_event(const l3gxxxx_t *dev)
+{
+    l3gxxxx_int_event_source_t source;
+
+    /* get the source of INT1 signal */
+    l3gxxxx_get_int_event_source(dev, &source);
+
+    /* display the event */
+    if (!source.active) {
+        return -1;
+    }
+    if (source.x_low) {
+        printf("x below ");
+    }
+    if (source.y_low) {
+        printf("y below ");
+    }
+    if (source.z_low) {
+        printf("z below ");
+    }
+    if (source.x_high) {
+        printf("x above ");
+    }
+    if (source.y_high) {
+        printf("y above ");
+    }
+    if (source.z_high) {
+        printf("z above ");
+    }
+    printf("defined thresholds: ");
+
+    return 0;
+}
+#endif
+
+int read_data(const l3gxxxx_t *dev)
+{
+#if IS_USED(MODULE_L3GXXXX_FIFO)
+
+    l3gxxxx_data_fifo_t data;
+
+    int num = l3gxxxx_read_fifo(dev, data);
+
+    printf("FIFO samples num=%d\n", num);
+    for (int i = 0; i < num; i++) {
+        printf("gyro [mdps] x: %+8" PRIi32 " y: %+8" PRIi32 " z: %+8" PRIi32 "\n",
+               data[i].x, data[i].y, data[i].z);
+    }
+
+    return num;
+
+#else /* IS_USED(MODULE_L3GXXXX_FIFO) */
+
+    l3gxxxx_data_t data;
+    l3gxxxx_read(dev, &data);
+
+    printf("gyro [mdps] x: %+8" PRIi32 " y: %+8" PRIi32 " z: %+8" PRIi32 "\n",
+           data.x, data.y, data.z);
+
+    return 1;
+
+#endif /* IS_USED(MODULE_L3GXXXX_FIFO) */
+}
+
+int main(void)
+{
+    l3gxxxx_t dev;
+
+    p_main = thread_getpid();
+
+    puts("L3GXXXX gyroscope driver test application\n");
+    puts("Initializing L3GXXXX sensor");
+
+    /* initialize the sensor with default configuration parameters */
+    if (l3gxxxx_init(&dev, &l3gxxxx_params[0]) == L3GXXXX_OK) {
+        puts("[OK]\n");
+    }
+    else {
+        puts("[Failed]");
+        return 1;
+    }
+
+#if IS_USED(MODULE_L3GXXXX_CONFIG)
+    /* select LPF2 and HPF, configure HPF and reset REFERENCE by dummy read */
+    int8_t ref;
+    l3gxxxx_select_output_filter(&dev, L3GXXXX_HPF_ONLY);
+    l3gxxxx_config_hpf(&dev, L3GXXXX_HPF_NORMAL, 0);
+    l3gxxxx_get_hpf_ref(&dev, &ref);
+#endif /* IS_USED(MODULE_L3GXXXX_CONFIG) */
+
+#if IS_USED(MODULE_L3GXXXX_FIFO)  && IS_USED(MODULE_L3GXXXX_CONFIG)
+    /* clear FIFO and activate FIFO Stream mode */
+    l3gxxxx_set_fifo_mode(&dev, L3GXXXX_BYPASS, 0);
+    l3gxxxx_set_fifo_mode(&dev, L3GXXXX_STREAM, 10);
+#endif
+
+#if IS_USED(MODULE_L3GXXXX_IRQ_DRDY)
+    /* init INT2/DRDY signal pin and enable the interrupt */
+    gpio_init_int(l3gxxxx_params[0].int2_pin, GPIO_IN, GPIO_RISING,
+                  l3gxxxx_int2_isr, 0);
+#endif /* IS_USED(L3GXXXX_IRQ_DRDY) */
+
+#if IS_USED(MODULE_L3GXXXX_IRQ_EVENT)
+    /* init INT1 signal pin and enable the interrupt */
+    gpio_init_int(l3gxxxx_params[0].int1_pin, GPIO_IN, GPIO_RISING,
+                  l3gxxxx_int1_isr, 0);
```

Comment:
In [`lis2dh12`](https://github.com/RIOT-OS/RIOT/blob/master/drivers/lis2dh12/lis2dh12.c#L473) I provided a wrapper function for this that blocks until an interrupt is received.

This was to prevent users from doing SPI directly inside the ISR to get the data.

Reply:
This is a different approach. In `lis2dh12` implementation the running thread calls `lis2dh12_wait_event` and blocks until the event happens and the driver takes care that the application doesn't access I2C or SPI in interrupt context.

At least I tried to document this [here](https://github.com/RIOT-OS/RIOT/blob/25dc4c4370d45a777cb4a4a7334337046b58954f/drivers/include/l3gxxxx.h#L780). In the L3Gxxxx driver, the application must use a mutex or messages in the ISR to inform the application thread that an interrupt has occurred. I believe that most of our drivers do it this way.

I can change this if you think it is more safe if the application does not access the sensor directly in the interrupt context.

Reply:
To be honest, I really like the idea that the driver ensures that an application doesn't access I2C or SPI in interrupt context. However, this has the disadvantage that the thread using the sensor cannot be used for anything else, which in turn increases the number of threads required for an application.

The advantage of the approach where the ISR is defined by the application and not by the driver is that the same thread can be used for other tasks if the ISR simply sets a flag used by the thread to determine when an event has occurred. This allows the thread to do other things between checking the flag.

Reply:
Which approach is preferred in RIOT to handle interrupts and the GPIO pins for these interrupts in drivers? After looking at a number of drivers, I am really unsure and the following two questions arose:

1. Who should be responsible to initialize the GPIO pin that is used for interrupts, e.g. when new sensor data are available? The driver or the application?
2. Should the ISR be part of the driver that uses a mutex and a blocking function call by the application? Or should the ISR be part of the application that is registered with the driver and called back when the interrupt occurs?

I have seen both approaches for both questions. We are talking about sensor drivers where interrupts are not used for internal functions but to indicate the occurrence of an event to the application.

Reply:
@benpicco Finally, I decided to change the interrupt handling to the approach as used in `lis2dh12` driver. So I label the PR as WIP until I have finished it.